File: Assets/Game/Scripts/UI/MobileControlSettingsUIControl.lua
local GameManager = Ballance2.Services.GameManager
local GameUIManager = GameManager.GetSystemService('GameUIManager') ---@type GameUIManager
local GameSettingsManager = Ballance2.Services.GameSettingsManager
local Image = UnityEngine.UI.Image
local EventTriggerListener = Ballance2.Services.InputManager.EventTriggerListener

---手机菜单控制菜单控制类
---@class MobileControlSettingsUIControl : GameLuaObjectHostClass
---@field ScrollView ScrollView
MobileControlSettingsUIControl = ClassicObject:extend()

function MobileControlSettingsUIControl:new() 
  self.items = {}
end
function MobileControlSettingsUIControl:Start() 
  local settings = GameSettingsManager.GetSettings('core')
  local page = GameUIManager:FindPage('PageSettingsControlsMobile')
  local controlKeypadSettting = ''

  ---@param index number
  ---@param item RectTransform
  self.ScrollView.updateFunc = function (index, item)
    local data = self.items[index + 1]
    if data then
      local EventTriggerListener = item:GetComponent(EventTriggerListener) ---@type EventTriggerListener
      local ImageBg = item:Find('ImageBg') 
      local ImageBgActive = item:Find('ImageBgActive') 
      local image = item:Find('Image'):GetComponent(Image) ---@type Image

      if controlKeypadSettting == data.name then
        ImageBg.gameObject:SetActive(false)
        ImageBgActive.gameObject:SetActive(false)
      else
        ImageBg.gameObject:SetActive(false)
        ImageBgActive.gameObject:SetActive(false)
      end

      image.sprite = data.image
      EventTriggerListener.onClick = function ()
        controlKeypadSettting = data.name
        settings:SetString('control.keypad', data.name)

        --更新列表
        self.ScrollView:UpdateData(false)

        --如果游戏正在运行中，则动态更换键盘
        if GamePlay.GamePlayManager then
          GameUI.GamePlayUI:ReBuildMobileKeyPad()
        end
      end
    end
  end
  self.ScrollView.itemCountFunc = function ()
    return #self.items
  end

  page.OnShow = function ()
    controlKeypadSettting = settings:GetString('control.keypad', 'BaseLeft')
    self:LoadList()
    self.ScrollView:UpdateData(false)
  end
end
function MobileControlSettingsUIControl:LoadList() 
  self.items = {}
  for key, value in pairs(KeypadUIManager.GetAllKeypad()) do
    if value then
      table.insert(self.items, {
        name = key,
        image = value.image,
      })
    end
  end
end

function CreateClass:MobileControlSettingsUIControl() 
  return MobileControlSettingsUIControl()
end

File: Assets/Game/Menu/Scripts/Control/MobileControlSettingsUIControl.cs
using System.Collections.Generic;
using Ballance2.Game;
using Ballance2.Game.GamePlay;
using Ballance2.Package;
using Ballance2.Services;
using Ballance2.Services.I18N;
using Ballance2.Services.InputManager;
using Ballance2.UI.Core;
using Ballance2.UI.Core.Controls;
using TMPro;
using UnityEngine;
using UnityEngine.UI;

namespace Ballance2.Menu
{
  public class MobileControlSettingsUIControl : MonoBehaviour {
    public ScrollView ScrollView;

    private class MobileControlItem {
      public string name;
      public string explain;
      public Sprite image;
    }
    private List<MobileControlItem> items = new List<MobileControlItem>();

    private void Start() {
      var settings = GameSettingsManager.GetSettings(GamePackageManager.SYSTEM_PACKAGE_NAME);
      var page = GameUIManager.Instance.FindPage("PageSettingsControls");
      var controlKeypadSettting = "";

      ScrollView.updateFunc = (index, item) => {
        var data = items[index];
        var Listener = item.GetComponent<EventTriggerListener>();
        var ImageBg = item.Find("ImageBg");
        var text = item.Find("Text").GetComponent<TMP_Text>();
        var ImageBgActive = item.Find("ImageBgActive");
        var image = item.Find("Image").GetComponent<Image>();

        if (controlKeypadSettting == data.name) {
          ImageBg.gameObject.SetActive(false);
          ImageBgActive.gameObject.SetActive(true);
        } else {
          ImageBg.gameObject.SetActive(true);
          ImageBgActive.gameObject.SetActive(false);
        }

        text.text = data.explain;
        image.sprite = data.image;

        Listener.onClick = (o) => {
          controlKeypadSettting = data.name;
          settings.SetString(SettingConstants.SettingsControlKeypad, data.name);

          //更新列表
          ScrollView.UpdateData(false);

          //如果游戏正在运行中，则动态更换键盘
          if (GamePlayUIControl.Instance != null)
            GamePlayUIControl.Instance.ReBuildMobileKeyPad();
        };
      };
      ScrollView.itemCountFunc = () => {
        return items.Count;
      };

      page.OnShow = (o) => {
        controlKeypadSettting = settings.GetString("control.keypad", "BaseCenter");
        LoadList();
        ScrollView.UpdateData(false);
      };
    }
    private void LoadList() {
      items.Clear();
      foreach (var item in KeypadUIManager.GetAllKeypad())
      {
        items.Add(new MobileControlItem() {
          name = item.Key,
          image = item.Value.image,
          explain = item.Value.explain,
        });
      }
    }
  }
}